// coc-html/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const coc_nvim_1 = require("coc.nvim");
const path_1 = tslib_1.__importDefault(require("path"));
function activate(context) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        let { subscriptions } = context;
        const config = coc_nvim_1.workspace.getConfiguration().get('html', {});
        const enable = config.enable;
        if (enable === false)
            return;
        const file = context.asAbsolutePath('lib/server/htmlServerMain.js');
        const selector = config.filetypes || ['html', 'handlebars', 'razor'];
        const embeddedLanguages = { css: true, javascript: true };
        let tagPaths = coc_nvim_1.workspace.getConfiguration('html').get('experimental.custom.tags', []);
        let attributePaths = coc_nvim_1.workspace.getConfiguration('html').get('experimental.custom.attributes', []);
        if (tagPaths && tagPaths.length > 0) {
            try {
                const workspaceRoot = coc_nvim_1.workspace.rootPath;
                tagPaths = tagPaths.map(d => {
                    return path_1.default.resolve(workspaceRoot, d);
                });
            }
            catch (err) {
                tagPaths = [];
            }
        }
        let serverOptions = {
            module: file,
            args: ['--node-ipc'],
            transport: coc_nvim_1.TransportKind.ipc,
            options: {
                cwd: coc_nvim_1.workspace.root,
                execArgv: config.execArgv || []
            }
        };
        let clientOptions = {
            documentSelector: selector,
            synchronize: {
                configurationSection: ['html', 'css', 'javascript']
            },
            outputChannelName: 'html',
            initializationOptions: {
                embeddedLanguages,
                tagPaths,
                attributePaths
            }
        };
        let client = new coc_nvim_1.LanguageClient('html', 'HTML language server', serverOptions, clientOptions);
        subscriptions.push(coc_nvim_1.services.registLanguageClient(client));
    });
}
exports.activate = activate;
